File: app/src/main/java/com/example/alarm/AlarmViewModel.java
package com.example.alarm;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Pair;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.alarm.model.Alarm;
import com.example.alarm.model.AlarmService;
import com.example.alarm.model.AlarmsListener;

import java.util.List;

import dagger.hilt.android.lifecycle.HiltViewModel;

import javax.inject.Inject;

@HiltViewModel
public class AlarmViewModel extends ViewModel {
    private static final String PREF_INTERVAL = "PREF_INTERVAL";
    private static final String PREF_WALLPAPER = "PREF_WALLPAPER";

    private final AlarmService alarmsService;
    private final SharedPreferences preferences;
    private final MutableLiveData<List<Alarm>> _alarms = new MutableLiveData<>();
    public LiveData<List<Alarm>> alarms = _alarms;

    private final MutableLiveData<Boolean> _initCompleted = new MutableLiveData<>();
    public LiveData<Boolean> getInitCompleted() {
        return _initCompleted;
    }

    private final AlarmsListener alarmsListener = _alarms::postValue;

    @Inject
    public AlarmViewModel(AlarmService alarmsService, SharedPreferences preferences) {
        this.alarmsService = alarmsService;
        this.preferences = preferences;

        alarmsService.initCompleted.observeForever(initCompleted -> {
            if (initCompleted) {
                _initCompleted.postValue(true);
            }
        });
        alarmsService.addListener(alarmsListener);
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        alarmsService.removeListener(alarmsListener);
    }

    public void updateEnabledAlarm(Alarm alarm, Boolean enabled, AlarmCallback callback) {
        alarmsService.updateEnabled(alarm.getId(), enabled);
        callback.onResult(true);
    }

    public void addAlarm(Alarm alarm, AlarmCallback callback) {
        boolean result = alarmsService.addAlarm(alarm);
        callback.onResult(result);
    }

    public void updateAlarm(Alarm alarmNew, AlarmCallback callback) {
        boolean result = alarmsService.updateAlarm(alarmNew);
        // Возвращаем результат через обратный вызов
        callback.onResult(result);
    }

    public void deleteAlarms(List<Alarm> alarmsToDelete, Context context) {
        alarmsService.deleteAlarms(alarmsToDelete, context);
    }

    public void getAndNotify() {
        alarmsService.getAlarms();
        alarmsService.notifyChanges();
    }

    public void getPreferencesWallpaperAndInterval(PreferenceCallback callback) {
        String wallpaper = preferences.getString(PREF_WALLPAPER, "");
        int interval = preferences.getInt(PREF_INTERVAL, 5);
        callback.onResult(new Pair<>(wallpaper, interval));
    }
}

File: app/src/main/java/com/example/alarm/AlarmViewModel.kt
package com.example.alarm

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.alarm.model.Alarm
import com.example.alarm.model.AlarmService
import com.example.alarm.model.AlarmsListener
import com.example.alarm.model.MyAlarmManager
import com.example.alarm.model.Settings
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

const val APP_PREFERENCES = "APP_PREFERENCES"
const val PREF_INTERVAL = "PREF_INTERVAL"

class AlarmViewModel(
    private val alarmsService: AlarmService
) : ViewModel() {

    private val _alarms = MutableLiveData<List<Alarm>>()
    val alarms: LiveData<List<Alarm>> = _alarms


    private val alarmsListener: AlarmsListener = {
        _alarms.value = it
    }
    init {
        alarmsService.addListener(alarmsListener)
    }


    override fun onCleared() {
        super.onCleared()
        alarmsService.removeListener(alarmsListener)
    }

    suspend fun updateEnabledAlarm(alarm: Alarm, enabled: Int, context: Context, idx: Int) = withContext(Dispatchers.Main) {
            if (alarm.enabled == 0) { //turn on
                val s = async(Dispatchers.IO) { alarmsService.getSettings() }
                MyAlarmManager(context, alarm, s.await()).startProcess()
            }
            else {
                MyAlarmManager(context, alarm, Settings(0)).endProcess()
            }
        withContext(Dispatchers.IO) {
            alarmsService.updateEnabled(alarm.id, enabled)
        }
        return@withContext idx
    }

    suspend fun addAlarm(alarm: Alarm, context: Context): Boolean = withContext(Dispatchers.Main) {
            if(withContext(Dispatchers.IO) { alarmsService.addAlarm(alarm)}) {
                val settings = async(Dispatchers.IO) { alarmsService.getSettings() }
                MyAlarmManager(context, alarm, settings.await()).startProcess()
                return@withContext true
            }
            else {
                return@withContext false
            }
    }

    suspend fun updateAlarm(alarmNew: Alarm, context: Context): Boolean = withContext(Dispatchers.Main) {
        if(withContext(Dispatchers.IO) { alarmsService.updateAlarm(alarmNew)}) {
            if (alarmNew.enabled == 1) {
                val settings = async(Dispatchers.IO) { alarmsService.getSettings() }
                MyAlarmManager(context, alarmNew, settings.await()).restartProcess()
            }
            return@withContext true
        }
        else {
            return@withContext false
        }
    }

    fun deleteAlarms(alarmsToDelete: List<Alarm>, context: Context?) {
        viewModelScope.launch {
            alarmsService.deleteAlarms(alarmsToDelete, context)
        }
    }
    fun getAndNotify() {
        viewModelScope.launch {
            alarmsService.getAlarms()
            alarmsService.notifyChanges()
        }
    }
    fun getPreferences(context: Context): Pair<String, Int>{
        val preferences: SharedPreferences = context.getSharedPreferences(APP_PREFERENCES, Context.MODE_PRIVATE)
        val wallpaper = preferences.getString(PREF_WALLPAPER, "")
        val interval: Int = preferences.getInt(PREF_INTERVAL, 5)
        return Pair(wallpaper!!, interval)
    }
}